server/commands/check.go

```diff
@@ -425,6 +425,10 @@ func (query *CheckQuery) resolveTupleToUserset(ctx context.Context, rc *resoluti
 			break // the user was resolved already, avoid launching extra lookups
 		}
 
+		if !tupleUtils.IsValidObject(tuple.GetUser()) && !tupleUtils.IsObjectRelation(tuple.GetUser()) {
```

Comment:
You should use `IsValidUser` here, otherwise you are ignoring valid relationship tuples, namely `user_id` and `*`

Reply:
What you need to be checking is that:
1) the user is of the correct user format
2) the user, if it has a type, is in the model
2) the user, if it has a relation, is in type definition of its type
2) the object, if it has a type, is in the model

Reply:
I'm not sure it makes sense for a `*` or `user_id` value to be acceptable for a tupleset relationship. A tupleset relationship should always be referencing an object, and not a user_id or `*`. Otherwise you can't resolve the computed_userset on the tupleset relation (e.g. `parent`).

Reply:
We want to ignore `user_id` here. Validation of stuff is done elsewhere. This is used to ensure that in a tuple to userset rewrite that we are only following usersets (and possibly `*`, but I haven't thought that far yet. It would be good to add a test of that case if we don't have it).

Reply:
This gate should happen after:
```go
userObj, userRel := tupleUtils.SplitObjectRelation(tuple.GetUser())
```
then
```go
if !tupleUtils.IsValidObject(userObj) {
    continue
}
```

The relation is already being handled/filtered appropriately.

Reply:
Consider the following model and tuples:

```
type group
  relations
    define viewer as self
    define member as self

type document
  relations
    define parent as self
    define viewer as viewer from parent
```

```
1. document:doc1, parent, group:eng#member
2. document:doc2, parent, folder:folder1
3. document:doc3, parent, myid
4. group:eng, viewer, jon
```

If the `parent` relationship of a document is defined as a tupleset relation, then only tuple 2 is valid (technically speaking). However, today you can write all three of these tuples. 

I think it's dangerous to allow tuple 1, because then what will happen here is the parent lookup will happen on `group:eng`. If `group:eng` had a viewer (e.g. jon) then that user will have been granted indirect access to the document unintentionally.

If we do what you're suggesting @elbuo8 we'll have
```
userObj, userRel := tupleUtils.SplitObjectRelation(tuple.GetUser())
if !tupleUtils.IsValidObject(userObject) {
  ...
}
// throw away userRel
// so userObject=group:eng & userObject=folder:folder1
```
The tupleset relation `parent` will be looked up on both of those, which will lead to the odd behavior above.

If we feel this is appropriate, then so be it. But that seems quite fragile and broken. @aaguiarz any input here.

Reply:
I don't think @elbuo8 said to throw away the `userRel` above. Just ensure that `userObj` is indeed an object.

Reply:
@craigpastro correct.